core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala

```diff
@@ -979,8 +976,9 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
       hadoopConf.set("mapred.output.compression.type", CompressionType.BLOCK.toString)
     }
 
-    // Use configured output committer if already set
-    if (conf.getOutputCommitter == null) {
+    // Use configured output committer if already set and speculation is not enabled.
+    val speculationEnabled = self.conf.getBoolean("spark.speculation", false)
+    if (speculationEnabled || conf.getOutputCommitter == null) {
```

Comment:
This is also somewhat tricky. Although rare, some output formats may set their own output committers. And if they do, they use the same configuration key (i.e. `mapred.output.committer.class`). One of the example is [Parquet] [1], which uses a specialized committer to write summary files. So unlike the case in Spark SQL, here we cannot distinguish a real user-defined committer (which is probably a direct committer), or a default committer bundled with the output format. I'm not quite sure forcing all output formats to use `FileOutputCommitter` when speculation is enabled is a good idea. On the other hand, I don't know any other output formats other than Parquet that use customized output committers, and it should be safe to assert that Spark users don't use `saveAsHadoopFile()` to write Parquet files.

[1]: https://github.com/apache/parquet-mr/blob/apache-parquet-1.7.0/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/mapred/DeprecatedParquetOutputFormat.java#L60